### PR TITLE
Python Requests with Stock Viewer

### PR DESCRIPTION
**Pull request by sohamb117**
- [x]  Workshop that meets all of the "Good workshops" guidelines outlined above (View below)

- [x]  1-2 minute video clip from your test club meeting showing what members built (View: https://cloud-k8zl09ofw.vercel.app/zoom_1.mp4 )

- [x]  All assets (images, PDFs, and any other binaries) must be hosted using the #cdn channel on the Hack Club Slack for posterity and to cut down on repository size.

- [x]  All workshops must be in English at native-level proficiency. Phrasing and grammar must be perfect.

- [x]  You must be either a registered Hack Club leader or an active member of the Slack.

Good workshops guidelines:
Help club members produce a compelling and real project in under 20 minutes

> We finished the workshop in around 15 minutes, not including technological difficulties with setting up the meeting. This is in my opinion a real project because it definitely teaches skills that are applicable to many other projects and can be developed in many ways to improve it.

Are very easy to hack on for long periods of time after completion

> Yes, this not only teaches the skills to make more projects like this but provides a starting point to build off of, and I provided a few examples of ways to build on this further. After the meeting I was even asked about some ways to extend the functionality, such as getting a list of the top stocks or the stocks with the most change.

Have instructive and compelling images, GIFs, and visuals

> I included some images to make it clearer. The demo is not that heavy on visuals, but it does have a few to help explain the concepts and detail the steps better.

Use repl.it or a similar service that allows members to begin coding with one click

> We used repl.it for the whole project and my demonstration was done on repl.it too.

Link unknown concepts to external sources.

> There's a list of links at the bottom to documentation and further reading, and links in text throughout to help provide more resources.

End with at least 3 demos of how the project can be expanded, with live links and code links for each demo.

> I included 3 demonstrations with short descriptions. These are done on repl.it with links to allow viewers to run and read my code.

This is a workshop on the Requests library in Python. It's intended to help people who already know some Python learn what HTTP requests are and how to connect APIs with their applications. I demoed this to my Hack Club in a mock meeting, since we plan to teach this once our Hack Club Python course starts, and we thought it went pretty well. Please make suggestions, I really want this to be good because I do think that the tutorials on the Requests library are pretty bland and I want this to be fun and informative.

### CODE REVIEW

**Review by zachlatta — CHANGES_REQUESTED** [review]
Hey @sohamb117, nice work on this! I did my first pass at https://youtu.be/7CC2XClln_8. Video is currently uploading. It's giving me an ETA of 2 hours until it's online.

Can you go through that and do a very detailed pass? I think this is a great start, but is about 30% of where we want it to be. We want to get to 100% and I think with a few big changes we can get there!

**Comment by sohamb117** [comment]
I made some changes, could you look this over and tell me what I need to fix up?

**Comment by sampoder** [comment]
I'll take a look at this for you :)

File: workshops/python_stock_viewer/README.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+There should be a key (I've hidden mine for security purposes - yours should be where the purple box is) but that's the key you'll need to use for authentication.
+
+Now let's set up the key in Python. The Finnhub API wants the key passed in through the headers, which are ways to send additional information. The Requests library allows you pass headers into the web request, which is how we'll authenticate. Let's define the headers.
+
+```python
+header = {'X-Finnhub-Token':'[TOKEN]'} 
+# Replace [TOKEN] with the token you got from Finnhub
+```
+How does this code work?
```
Comment (sampoder) [inline workshops/python_stock_viewer/README.md:68]:
```suggestion
How does this code work?  

```

**Review by sampoder — APPROVED** [review]
Sick workshop, love it 🚀 

@zachlatta I think you can take it from here? Not sure what to do from here :D

**Comment by sampoder** [comment]
Wait shoot, I just realized we may want to bulk this workshop out. Sorry for backtracking but could we add the input functionality into the workshop and create another demo. Then I think it'd become a real product

Sorry sorry sorry, ik this is a really ******* thing to do sorry

**Review by sampoder — CHANGES_REQUESTED** [review]
see above, sorry again

**Comment by sampoder** [comment]
the workshop is still super well done, sorry

**Comment by sohamb117** [comment]
Oh, sure. Sorry, I had it in originally but I realized that it was getting long in terms of time. Not that hard to add it in so I'll do that I guess.

**Review by sampoder — APPROVED** [review]
I think it's now ready, yay!!!! 🎉

**Comment by sohamb117** [comment]
Thanks for the help!

**Comment by zachlatta** [comment]
Amazing! I will do my best to do a final review this week. Things are really crazy right now.

**Comment by sohamb117** [comment]
Hey, @zachlatta, can I get an update sometime soon? I know it's been busy, but I'd like this reviewed soon. Could you look it over when you get the chance?

**Comment by zachlatta** [comment]
Hey Sohum, you bet. I appreciate your patience. We’re in the midst of
moving HQ right now and are moving into our new place tomorrow afternoon,
hence the delay.

I should be able to get to by mid or late this week.

On Oct 4, 2020, at 13:55, Soham Bhattacharya <notifications@github.com>
wrote:

﻿

Hey, @zachlatta <https://github.com/zachlatta>, can I get an update
sometime soon? I know it's been busy, but I'd like this reviewed soon.
Could you look it over when you get the chance?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/hackclub/hackclub/pull/1304#issuecomment-703291618>, or
unsubscribe

**Comment by sohamb117** [comment]
Thanks for the update!

**Comment by sohamb117** [comment]
@zachlatta Any chance I can get another update? I know you're busy, but I'd really appreciate it if you could review this soon. Thanks!

**Comment by MatthewStanciu** [comment]
@zachlatta if it’s okay with you, I’m going to take this off your plate and review it tomorrow

File: workshops/python_stock_viewer/README.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+![DEMO](https://cloud-2v062st9k.vercel.app/screen-recording-online-video-cu.gif)
+
+
+[See this in action](https://repl.it/@sohamb117/demo2)
+
+
+## What are HTTP Requests?
+
+HTTP requests are a type of request a program can send to a server to get data back. You might use requests to get weather information, or information from a database, or a video or audio file. There are many types of requests one can use, but the type we'll be covering here are GET requests, which send a request to the server or API with arguments in either the headers of the request or the URL, and expect some response data back. Google, Amazon, pretty much every site amd app available uses requests to get data from the Internet.
```
Comment (bezlin6mechminerz) [inline workshops/python_stock_viewer/README.md:13]:
```suggestion
HTTP requests are a type of request a program can send to a server to get data back. You might use requests to get weather information, or information from a database, or a video or audio file. There are many types of requests one can use, but the type we'll be covering here are GET requests, which send a request to the server or API with arguments in either the headers of the request or the URL, and expect some response data back. Google, Amazon, pretty much every site and app available uses requests to get data from the Internet.
```

Comment (bezlin6mechminerz) [reply workshops/python_stock_viewer/README.md:13]: Do you mean 'and' here?

Comment (sohamb117) [reply workshops/python_stock_viewer/README.md:13]: Oh yeah! Thanks!

File: workshops/python_stock_viewer/README.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
+# Web Requests in Python - With a Stock Visualizer!
```
Comment (lachlanjc) [inline workshops/python_stock_viewer/README.md:1]: Needs top metadata, as in the guidelines

File: workshops/python_stock_viewer/README.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+Here are some things to read to learn more about what we did here:
+
+* https://requests.readthedocs.io/en/master/user/quickstart/#make-a-request
+* https://www.tutorialspoint.com/http/http_requests.htm
+* https://docs.python.org/3/library/json.html
+* https://www.howtogeek.com/343877/what-is-an-api/
+
+
+[Workshop Demo Video](https://cloud-k8zl09ofw.vercel.app/zoom_1.mp4)
```
Comment (lachlanjc) [inline workshops/python_stock_viewer/README.md:216]: Don’t link this in the workshop—this is just for your PR submission
```suggestion
```

**Review by lachlanjc — CHANGES_REQUESTED** [review]
I love where this is going, & your explanations are super solid! However, this workshop is not nearly complete enough—the end result needs to be a bunch more than 9 lines of code. Even if you don’t change the functionality in a huge way, you could do things like formatting the currency, showing the company name info/other stats like percentage change since `pc`, making a tiny ASCII chart, etc. Or you could use this to server-render a little website with the info & a form? That’d be awesome. Whatever direction you take it, I don’t have strong feelings, but this currently just isn’t long/complete/thorough enough. Super appreciate your submission & we’d love to review a revised version that walks through more extended code—you’ve got an amazing start here.

**Pull request closed without merging**